tools/reviewer/web/src/app/dashboard/diff/diff-header/add-user-dialog/add-user-dialog.component.ts

```diff
@@ -0,0 +1,51 @@
+import { Component, HostListener } from '@angular/core';
+import { MatDialogRef } from '@angular/material';
+
+@Component({
+  selector: 'add-user-dialog',
+  templateUrl: './add-user-dialog.component.html',
+  styleUrls: ['./add-user-dialog.component.scss'],
+})
+export class AddUserDialogComponent {
+  isEmailInvalid: boolean = false;
+  email: string;
+
+  constructor(public dialogRef: MatDialogRef<AddUserDialogComponent>) { }
+
+  add(): void {
+    if (this.isEmailValid(this.email)) {
+      this.dialogRef.close(this.email);
+    } else {
+      this.isEmailInvalid = true;
+    }
+  }
+
+  close(): void {
+    this.dialogRef.close();
+  }
+
+  isEmailValid(email: string): boolean {
+    if (!email) {
+      return false;
+    }
+
+    // Example: my_super-name@flip-flop.com
+    const emailValidation: RegExp = /^[\w-.]+@[\w-]+\.[\w]+$/;
```

Comment:
Where is this regex from?
This one seems more complicated: http://emailregex.com/
Are you sure we're covering all cases?

Reply:
Is this regex being used just to let the user know if the email address they entered is _likely_ to be valid, or is it being used to _force_ the user the enter a valid address?

If it's the latter, this is bad practice for reasons described in https://stackoverflow.com/a/202528/2252930.

But if it's the former, then please, do carry on! :)

Reply:
Force the user to enter a valid email address.
We're not planning to send emails to this address (at least not without validating it as is mentioned in the link), but if we can prevent at least _some_ manual entry mistakes, I think we should.
If it ends up being a problem that people are entering wrong emails, we'll think of a more elaborate solution.

Reply:
@oferb Sorry, I'm a bit lost.

Are you advocating forcing a user to enter an email address that matches the regex, no exceptions?

Or are you suggesting that, like most other sites (that I've personally used), it should just act as a suggestion for the user, allowing them to enter their email address even if it doesn't match the regex exactly?

Reply:
Yeah, I know there're much more difficult email regexes, but to be honest, I made the validation only to be sure that email has basic email structure and we can take username from it.

Reply:
@webdevelopland That sounds fair to me. :+1:

Reply:
Why would a user want to enter an email that the regex rejects?
That would imply the regex is not good, and in that case, we should use a different one :)

Reply:
@oferb As you kind of implied, a user would expect a website to allow them to enter their email address even if it deviates from whatever preconceived notion of an "email address" the website may have, but is otherwise a perfectly valid email address. In other words, the way I see it, the website should not use a regex or whatever as the absolute validator; it should instead use the regex to _guide_ the user on the UI side, and then after the user has hit Enter, the website should send an email to that email address so that the user can confirm that they've entered the correct address.

I was burned by something kind of similar in the past, where I wasn't able to use my email address to create an account on a website (which I forget the name of now) because my email address was considered too long, even though it was legitimate. Now, I know, it's not the same problem as having an incorrect regex, but I think it still shows that, ultimately, the final validation _shouldn't_ happen on the website itself, but rather by sending a confirmation email to the entered email address.

Reply:
Got it. I didn't know email format was so ill-defined. The link I shared mentions `RFC 5322 Official Standard`. I thought there's an accepted standard for email addresses, but I guess I'm wrong. In that case, let's indeed stay with a minimally-restrictive regex.

Reply:
> The link I shared mentions RFC 5322 Official Standard. I thought there's an accepted standard for email addresses.

And I think you're right that it is the standard!

But alas, according to your link, "[there is no perfect email regex](http://www.regular-expressions.info/email.html)", which matches up with what I remember having read about email regexes in the past in that not _all_ issued email addresses follow the standard, so trying to validate email addresses against it is probably not the best idea. :)